11_Robots/ghost_tank/src/tank_robot_plugin.cpp

```diff
@@ -390,23 +432,62 @@ void TankRobotPlugin::teleop(double current_time)
 //   }
 // }
 
-void TankRobotPlugin::worldOdometryUpdateCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
+void TankRobotPlugin::imuUpdateCallback(const sensor_msgs::msg::Imu::SharedPtr msg)
 {
-  double theta = ghost_util::quaternionToYawRad(
-    msg->pose.pose.orientation.w,
-    msg->pose.pose.orientation.x,
-    msg->pose.pose.orientation.y,
-    msg->pose.pose.orientation.z);
-  m_tank_model_ptr->setWorldPose(msg->pose.pose.position.x, msg->pose.pose.position.y, theta);
-  m_tank_model_ptr->setWorldTwist(
-    msg->twist.twist.linear.x,
-    msg->twist.twist.linear.y,
-    msg->twist.twist.angular.z);
+  // TODO do we even need this anymore
+  m_imu_yaw = ghost_util::quaternionToYawRad(
+    msg->orientation.w,
+    msg->orientation.x,
+    msg->orientation.y,
+    msg->orientation.z);
 }
 
+//void TankRobotPlugin::worldOdometryUpdateCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
+//{
+//if (!m_use_backup_estimator) {
+//  double theta = ghost_util::quaternionToYawRad(
+//    msg->pose.pose.orientation.w,
+//    msg->pose.pose.orientation.x,
+//    msg->pose.pose.orientation.y,
+//    msg->pose.pose.orientation.z);
+//  m_tank_model_ptr->setWorldPose(msg->pose.pose.position.x, msg->pose.pose.position.y, theta);
+//  m_tank_model_ptr->setWorldTwist(
+//    msg->twist.twist.linear.x,
+//    msg->twist.twist.linear.y,
+//    msg->twist.twist.angular.z);
+//}
+//}
+
+//void TankRobotPlugin::worldOdometryUpdateCallbackBackup(
+//  const nav_msgs::msg::Odometry::SharedPtr msg)
+//{
+//if (m_use_backup_estimator) {
+//  double theta = ghost_util::quaternionToYawRad(
+//    msg->pose.pose.orientation.w,
+//    msg->pose.pose.orientation.x,
+//    msg->pose.pose.orientation.y,
+//    msg->pose.pose.orientation.z);
+//  m_tank_model_ptr->setWorldPose(msg->pose.pose.position.x, msg->pose.pose.position.y, theta);
+//  m_tank_model_ptr->setWorldTwist(
+//    msg->twist.twist.linear.x,
+//    msg->twist.twist.linear.y,
+//    msg->twist.twist.angular.z);
+//}
+//}
+
+//void TankRobotPlugin::publishBaseTwist()
+//{
+//geometry_msgs::msg::Twist msg{};
+//auto base_vel_cmd = m_tank_model_ptr->getBaseVelocityCommand();
+//msg.linear.x = base_vel_cmd.x();
+//msg.linear.y = base_vel_cmd.y();
+//msg.angular.z = base_vel_cmd.z();
+//m_base_twist_cmd_pub->publish(msg);
+//}
+
```

Comment:
do i really have to publish twist?? i can tho, it should be pretty trivial by diffing the last two odom vals.